plugins/destination/postgresql/client/migrate.go

```diff
@@ -48,165 +167,193 @@ func (c *Client) Migrate(ctx context.Context, tables schema.Tables) error {
+		// the user will have to drop and recreate the table
+		if pgErr.Code == "23502" {
+			return fmt.Errorf("column %s.%s contains null values, please drop the table manually via 'drop table %s'", tableName, column.Name, tableName)
+		}
+		// this is a weird CockroachDB error code that means we need to retry.
+		if pgErr.Code != "55000" {
+			return fmt.Errorf("failed to add column %s.%s with pgerror %s: %w", tableName, column.Name, pgErrToStr(pgErr), err)
+		}
+		if _, err := c.conn.Exec(ctx, sql); err != nil {
+			return fmt.Errorf("failed to retry adding column %s.%s with pgerror %s: %w", tableName, column.Name, pgErrToStr(pgErr), err)
 		}
 	}
-	return stalePks
+	return nil
 }
 
-func (c *Client) autoMigrateTable(ctx context.Context, table *schema.Table) error {
-	var err error
-	var pgColumns *pgTableColumns
-	var pgPKs map[string]bool
-
-	// create the new column as it doesn't exist
-	tableName := pgx.Identifier{table.Name}.Sanitize()
-	if pgColumns, err = c.getPgTableColumns(ctx, table.Name); err != nil {
-		return fmt.Errorf("failed to get table %s columns types: %w", table.Name, err)
+func (c *Client) alterColumnPg(ctx context.Context, tableName string, column schema.Column) error {
+	columnName := pgx.Identifier{column.Name}.Sanitize()
+	columnType := c.SchemaTypeToPg(column.Type)
+	// try alter column in place first
+	sql := "alter table " + tableName + " alter column " + columnName + " type " + columnType
+	if column.CreationOptions.NotNull {
+		sql += " set not null"
 	}
-
-	if pgPKs, err = c.getPgTablePrimaryKeys(ctx, table.Name); err != nil {
-		return fmt.Errorf("failed to get table %s primary key columns: %w", table.Name, err)
+	if _, err := c.conn.Exec(ctx, sql); err != nil {
+		c.logger.Warn().Err(err).Str("table", tableName).Str("column", column.Name).Msg("Column type changed in place failed.")
 	}
-
-	stalePks := c.getStalePks(pgPKs, table)
-	constraintName := pgx.Identifier{table.Name + "_cqpk"}.Sanitize()
-	dropConstraintSQL := "alter table " + tableName + " drop constraint if exists " + constraintName
-	if len(stalePks) > 0 {
-		message := "the following primary keys were removed from the schema %q for table %q.\nYou can migrate the table manually by running:\n%s"
-		sep := strings.Repeat("-", len(dropConstraintSQL)+1)
-		query := fmt.Sprintf("%s\n%s;\n%s\n%s", sep, dropConstraintSQL, getDropNotNullQuery(table, stalePks), sep)
-		return fmt.Errorf(message, stalePks, table.Name, query)
+	tx, err := c.conn.BeginTx(ctx, pgx.TxOptions{})
+	if err != nil {
+		return fmt.Errorf("failed to start alter column transaction: %w", err)
 	}
-
-	reCreatePrimaryKeys := false
-
-	for _, col := range table.Columns {
-		columnName := pgx.Identifier{col.Name}.Sanitize()
-		columnType := c.SchemaTypeToPg(col.Type)
-		if columnType == "" {
-			c.logger.Warn().Str("table", table.Name).Str("column", col.Name).Msg("Column type not supported, skipping")
-			continue
+	defer func() {
+		if err := tx.Rollback(ctx); err != nil {
```

Comment:
I think Rollback will always fail if the Commit succeeds as it will try to rollback with a closed connection

Reply:
It is fine to call rollback after commit in pgx as long as we check if error is closed and then there is no need to log it. This way we ensure there is cleanup.

Reply:
👍 I'm good with always calling rollback, just at the moment it prints errors when it shouldn't so we should add a check there on the error type